[#536] Add support for custom private key file for public-key export and public-key export-address commands

diega opened the pull request:
## PR description
This PR adds `--node-private-key-file` option to `public-key export-address` and `public-key export` commands.

## Fixed Issue(s)
Fixes #536

## Changelog

- [ ] Add support for custom private key file for public-key export and public-key export-address commands

atoulme commented on besu/src/test/java/org/hyperledger/besu/cli/PublicKeySubCommandTest.java:
```diff
@@ -141,6 +177,30 @@ public void callingPublicKeyExportSubCommandWithFilePathMustWritePublicKeyInThis
   }
 
+  @Test
+  public void callingPublicKeyExportSubCommandWithPrivateKeyFileMustWriteKeyToStandardOutput()
```
What is the behavior if the file doesn't exist or cannot be read?

diega replied:
It throws an exception. I just added a test that reflects that behavior

macfarla commented:
org.hyperledger.besu.Besu public-key export-address --node-private-key-file=/nope/issue-536.key
picocli.CommandLine$ExecutionException: Error while running command (org.hyperledger.besu.cli.subcommands.PublicKeySubCommand$AddressSubCommand@5f3b9c57): java.lang.IllegalArgumentException: Cannot store generated private key.
	at picocli.CommandLine.executeUserObject(CommandLine.java:1948)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:2172)
	at picocli.CommandLine.parseWithHandlers(CommandLine.java:2550)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:53)
	at org.hyperledger.besu.cli.util.ConfigOptionSearchAndRunHandler.handle(ConfigOptionSearchAndRunHandler.java:31)
	at picocli.CommandLine$AbstractParseResultHandler.handleParseResult(CommandLine.java:2172)
	at picocli.CommandLine.parseWithHandlers(CommandLine.java:2550)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1363)
	at org.hyperledger.besu.cli.BesuCommand.parse(BesuCommand.java:1192)
	at org.hyperledger.besu.Besu.main(Besu.java:50)
Caused by: java.lang.IllegalArgumentException: Cannot store generated private key.
	at org.hyperledger.besu.crypto.KeyPairUtil.storeKeyFile(KeyPairUtil.java:91)
	at org.hyperledger.besu.crypto.KeyPairUtil.loadKeyPair(KeyPairUtil.java:73)
	at org.hyperledger.besu.cli.BesuCommand.loadKeyPair(BesuCommand.java:1329)
	at org.hyperledger.besu.cli.subcommands.PublicKeySubCommand$AddressSubCommand.run(PublicKeySubCommand.java:168)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
	... 13 more

Process finished with exit code 1

macfarla commented on besu/src/main/java/org/hyperledger/besu/cli/options/stable/NodePrivateKeyFileOption.java:
```diff
@@ -0,0 +1,39 @@
+/*
+ * Copyright ConsenSys AG.
```
nit: new files should have * Copyright contributors to Hyperledger Besu  https://wiki.hyperledger.org/display/TSC/Copyright+and+License+Policy

macfarla reviewed (changes requested):
When testing this locally, if I specify a non-existent path, I get an exception.
If I specify an existent path but a file that doesn't exist, a `key` file is generated in the supplied path - see comments

Would be nice to add tests for this.

diega commented:
Hello @macfarla, thanks for the review. I think I addressed your comments.
I've added a check for the existence of the `--node-private-key-file` so I can throw a  `ParameterException` as you suggested. With this I think there is no need to cover the case for the `key` file creation for the case when the file doesn't exists but the parent directory does.
I've also added a test for when the file doesn't contain a valid private key.

macfarla commented on besu/src/main/java/org/hyperledger/besu/cli/subcommands/PublicKeySubCommand.java:
```diff
@@ -153,34 +124,52 @@ private void outputPublicKey(final NodeKey nodeKey) {
+        throw new CommandLine.ParameterException(
+            spec.commandLine(), "Private key cannot be loaded from file", e);
+      }
+      String output = outputFunction.apply(keyPair);
```
variable can be final

macfarla commented on besu/src/main/java/org/hyperledger/besu/cli/subcommands/PublicKeySubCommand.java:
```diff
@@ -153,34 +124,52 @@ private void outputPublicKey(final NodeKey nodeKey) {
-      // otherwise print to standard output.
-      if (addressExportFile != null) {
-        final Path path = addressExportFile.toPath();
+      File nodePrivateKeyFile = nodePrivateKeyFileOption.getNodePrivateKeyFile();
```
final

macfarla commented on besu/src/main/java/org/hyperledger/besu/cli/options/stable/NodePrivateKeyFileOption.java:
```diff
@@ -0,0 +1,39 @@
+      names = {"--node-private-key-file"},
+      paramLabel = MANDATORY_PATH_FORMAT_HELP,
+      description =
+          "The node's private key file (default: a file named \"key\" in the Besu data folder)")
```
nit: s/folder/directory

macfarla reviewed (approved):
couple of small comments, overall LGTM!

macfarla commented:
Also @diega your changelog entry is good - can you add it to the CHANGELOG.md file under "Additions"
> Add support for custom private key file for public-key export and public-key export-address commands

macfarla commented:
All merged in - thanks @diega for your contribution!
